consensus/state.go

```diff
@@ -2066,6 +2020,19 @@ func (cs *State) recordMetrics(height int64, block *types.Block) {
 	cs.metrics.CommittedHeight.Set(float64(block.Height))
 }
 
+// isReadyToPrecommit calculates if the process has waited at least 6 seconds
+// from their start time before they can vote
+// If the cs.config.DelayedPrecommitTimeout is set to 0, no precommit wait is done.
+func (cs *State) isReadyToPrecommit() (bool, time.Duration) {
+	if cs.config.DelayedPrecommitTimeout == 0 {
+		// setting 0 as a special case not to reschedule the pre-commit
+		return true, 0
+	}
+	precommitVoteTime := cs.rs.StartTime.Add(cs.config.DelayedPrecommitTimeout)
```

Comment:
how do we determine round start time?

Reply:
for rounds, it doesn't really matter. In fact, the way I understand it: Delayed precommits allows pacing the network to a certain block time for small blocks. For big blocks, they wouldn't benefit from it since they're already taking the whole period to propagate/process. So, if there is a new round, we wouldn't delay the precommit ever since we're already late and want to commit a block.

What delayed precommits allow:

- having a huge timeout propose
- a miniscule timeout commit that can go down to 1ms. In all cases, we take ~2s to be able to propose a block (prepare proposal + propagation block encoding + proof generation).

So, we don't need to count for round.